### ISSUE TITLE
Block installation of specific apps

### ISSUE THREAD

**Issue by quenenni** [nextcloud/server#21057]
**Is your feature request related to a problem? Please describe.**

Before, we were only offering a list of apps we trusted, but we want to be more open and also give a chance to new apps (and tbh, that was a lot of work too to maintain).
That's why we decide to let users choose between all the apps available.

Unfortunately, some apps are weaker than others and just a good description will bring lots of people to try it (trusting the apps listed in their apps page).

I have that particular problem with the app "Group Folders", that I must say offers something very nice.
But also, I have regularly big problems due to this app, problems that are always a pain to debug and cost me way too much time to fix (plus often, all the data are unavailable for an amount of time).

I'd be happy to make it available in the future, but I need to be sure the app is good enough before I let it be installed.

**Describe the solution you'd like**
Couldn't we have somewhere (in config.php file or via the parameters page), a way to select apps we don't want to be installed or used in our nextcloud?

**Describe alternatives you've considered**
I don't see alternative atm.

**Comment by kesselb** [comment]
It's possible to specify a custom app store url. Mirror apps.json and categories.json to your own server, setup a job to regularly update the files from the official app store and remove apps you don't want.

**Comment by quenenni** [comment]
@kesselb  Thanks a lot for your help.

I see waht you mean, but is it possible to manage a custom app you have at the same time as what you propose?
Isn't that a bit complexe for what I would like to achieve?

**Comment by kesselb** [comment]
> I see waht you mean, but is it possible to manage a custom app you have at the same time as what you propose?

Can you rephrase the question? I don't get it.

> Isn't that a bit complexe for what I would like to achieve?

It's actually something like download apps.json, parse the json, remove app if on your blacklist, save again as apps.json. Point your instances to your app store url.

**Comment by quenenni** [comment]
We planned to develop one or 2 custom apps for our nextcloud, and so use the custom app system.
It's not yet done, still in consideration / discussion.

As I don't know yet how that custom app system works, I was wondering if it will be easy to mix your own custom apps and the system you're explaining.

**Comment by benks-io** [comment]
As described in #24815 the hoster of a next cloud Instance should be able to hide certain apps from the administrator, that only has access to the GUI. I consider this a common requirement and would not suggest admins to set up their own app store to achieve that goal, as this is a lot of overhead.

I propose to manage visible apps booth through commands and lists (e.g. in the database):

a black list (certain apps cannot be installed, every new app can still be installed)
a white list (only marked apps can be installed at all, new apps will only become available when white listed)

To be more specific - I propose:

- occ app:availability:default hide
- occ app:availability:default show

- occ app:availability:hide $APPNAME
- occ app:availability:hide_all

- occ app:availability:show $APPNAME
- occ app:availability:show_all

and to show only featured apps:

- occ app:availablity:default show_only_featured_apps

The command
- occ app:availability:show_status $APPNAME

Should show something like:
"$APPNAME is blacklisted"
"$APPNAME is whitelisted"
"$APPNAME has no status and will be shown by default"
"$APPNAME has no status and will not be shown by default"


The solution should be prepared to add the feature later to block or allow certain versions of an App.

**Comment by quenenni** [comment]
I'm sorry to come back on this post, but I think it becomes more and more important to have this.

Here are another cases of problems we had lately...

In our clouds, we configure a shared Lool and a shared onlyoffice for every cloud instances.
But some users just install themselves the opendocument server and try to configure it in their own cloud.

On the next cloud update, all the clouds that had that crashed during the update process because of this app.

Another user of a cloud just decided to install more than 80 apps -> Cloud down because some of the apps were really a bad idea to install.

**Comment by benks-io** [comment]
Yes to separate the responsibilities of hosting (stability, availability) and NC administration (user management, configuration, theming etc.) seems to me increasingly important, too.

**Comment by summersab** [comment]
My idea in #21566 was to add a little bit of code so that black/whitelists could be added to config.php. It would be pretty simple to do with about a dozen lines of code.

**Comment by benks-io** [comment]
@summersab As a server admin I would prefare not to be forced to set up an appstore or to use docker but to use a built in way.

**Comment by szaimen** [comment]
To me #21566 sounds like a good proposal for anyone interested in implementing this 👍

**Comment by summersab** [comment]
@benks-io You wouldn't really be setting up an app store but rather creating a "dummy" app store. Basically, you're just downloading the app store listings from Nextcloud's servers, filtering out what you don't want, storing the listings locally, and pointing your `config.php` to the listing files. You don't need to host all of the apps and files locally - it's not that extensive.

To make this easier, I came up with a script that implements what I just described (@quenenni and @szaimen - you may be interested). This basically does the same thing as my feature request in #21566, only this doesn't require modifications the core Nextcloud libraries and keeps things as native as possible. It could probably be turned into an app, but I'm not familiar with writing code for Nextcloud. Installation instructions for the script are in the header comments:
https://gist.github.com/summersab/363c753c4936f0bc563ea88a85de6241 

If you have questions or need help, feel free to ask. It might be better to post a comment on my Gist instead of replying here. Also, if you need more help like this, I happen to be a freelance PHP developer. Like I said, I don't currently have experience writing code for Nextcloud, but I've been wanting an excuse to learn. I'd be happy to cut my rates for any work I do while I get familiar with Nextcloud's code.

**Comment by szaimen** [comment]
Lets us track this in https://github.com/nextcloud/server/issues/21566
